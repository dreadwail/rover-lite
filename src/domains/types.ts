import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
