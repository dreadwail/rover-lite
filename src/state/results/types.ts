export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
