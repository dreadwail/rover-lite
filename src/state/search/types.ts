import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
