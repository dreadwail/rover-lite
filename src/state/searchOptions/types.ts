export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
