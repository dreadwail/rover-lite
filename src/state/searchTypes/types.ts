export type SearchType =
  | 'overnight-boarding'
  | 'overnight-traveling'
  | 'drop-in'
  | 'doggy-day-care'
  | 'dog-walking';

export type Endpoints = { [T in SearchType]?: string };

export interface SearchTypes {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly endpoints: Endpoints;
}
