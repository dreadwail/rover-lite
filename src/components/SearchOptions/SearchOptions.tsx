import React, { Component } from 'react';

import { SearchInputs } from '../../state/search/types';
import { OptionsByCategoryId, OptionCategory } from '../../state/searchOptions/types';
import { SearchType } from '../../state/searchTypes/types';
import ReloadableFailure from '../ReloadableFailure';
import SearchOptionsCategory from '../SearchOptionsCategory';
import SearchOptionsSet from '../SearchOptionsSet';
import Spinner from '../Spinner';

export interface SearchOptionsProps {
  readonly searchType: SearchType;
  readonly isFetching: boolean;
  readonly isSearchTypesLoaded: boolean;
  readonly hasFailed: boolean;
  readonly optionsByCategoryId: OptionsByCategoryId;
  readonly rankedCategories: OptionCategory[];
  readonly inputs: SearchInputs;
  fetchSearchOptions(searchType: SearchType): void;
  onSearchFieldChanged(key: string, value: any): void;
}

export default class SearchOptions extends Component<SearchOptionsProps> {
  public componentDidMount() {
    const { isSearchTypesLoaded, searchType, fetchSearchOptions } = this.props;
    if (isSearchTypesLoaded) {
      fetchSearchOptions(searchType);
    }
  }

  public componentDidUpdate(previousProps: SearchOptionsProps) {
    const {
      searchType: oldSearchType,
      isSearchTypesLoaded: oldIsSearchTypesLoaded,
    } = previousProps;
    const { isSearchTypesLoaded, fetchSearchOptions, searchType } = this.props;
    const searchTypesLoaded = oldIsSearchTypesLoaded === false && isSearchTypesLoaded === true;
    const searchTypeChanged = oldSearchType !== searchType;
    const requiresRefetch = searchTypesLoaded || searchTypeChanged;
    if (isSearchTypesLoaded && requiresRefetch) {
      fetchSearchOptions(searchType);
    }
  }

  public render() {
    const {
      rankedCategories,
      isFetching,
      hasFailed,
      optionsByCategoryId,
      inputs,
      onSearchFieldChanged,
    } = this.props;

    if (isFetching) {
      return <Spinner />;
    }

    if (hasFailed) {
      return <ReloadableFailure />;
    }

    return (
      <>
        {rankedCategories.map(category => {
          const options = optionsByCategoryId[category.categoryId];
          return (
            <SearchOptionsCategory key={category.categoryId} {...category}>
              <SearchOptionsSet
                options={options}
                inputs={inputs}
                onSearchFieldChanged={onSearchFieldChanged}
              />
            </SearchOptionsCategory>
          );
        })}
      </>
    );
  }
}
