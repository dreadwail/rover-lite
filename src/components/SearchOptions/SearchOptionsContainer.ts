import { connect, MapStateToProps } from 'react-redux';

import { searchFieldChanged } from '../../domains/search/actions';
import { getInputs, getSearchType } from '../../domains/search/selectors';
import { fetchSearchOptions } from '../../domains/searchOptions/actions';
import {
  getOptionsByCategoryId,
  getRankedCategories,
  hasFailed,
  isFetching,
} from '../../domains/searchOptions/selectors';
import { isSearchTypesLoaded } from '../../domains/searchTypes/selectors';
import { MapDispatchToProps } from '../../domains/types';
import { RoverSearchState } from '../../state/types';

import SearchOptions, { SearchOptionsProps } from './SearchOptions';

export type OwnProps = {};
export type MappedStateProps = Pick<
  SearchOptionsProps,
  | 'searchType'
  | 'isFetching'
  | 'isSearchTypesLoaded'
  | 'hasFailed'
  | 'optionsByCategoryId'
  | 'rankedCategories'
  | 'inputs'
>;
export type MappedDispatchProps = Pick<
  SearchOptionsProps,
  'fetchSearchOptions' | 'onSearchFieldChanged'
>;

export const mapStateToProps: MapStateToProps<
  MappedStateProps,
  OwnProps,
  RoverSearchState
> = state => ({
  searchType: getSearchType(state),
  isFetching: isFetching(state),
  isSearchTypesLoaded: isSearchTypesLoaded(state),
  hasFailed: hasFailed(state),
  optionsByCategoryId: getOptionsByCategoryId(state),
  rankedCategories: getRankedCategories(state),
  inputs: getInputs(state),
});

export const mapDispatchToProps: MapDispatchToProps<MappedDispatchProps, OwnProps> = dispatch => ({
  fetchSearchOptions: async searchType => {
    await dispatch(fetchSearchOptions(searchType));
  },
  onSearchFieldChanged: (key, value) => {
    dispatch(searchFieldChanged(key, value));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchOptions);
