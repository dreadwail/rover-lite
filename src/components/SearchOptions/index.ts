export * from './SearchOptions';
export { default } from './SearchOptionsContainer';
