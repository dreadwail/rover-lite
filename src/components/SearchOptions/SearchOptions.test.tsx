import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import { OptionCategory } from '../../state/searchOptions/types';

import SearchOptions, { SearchOptionsProps } from './SearchOptions';

describe('SearchOptions', () => {
  let props: SearchOptionsProps;
  let rendered: ShallowWrapper<SearchOptionsProps>;

  beforeEach(() => {
    props = {
      searchType: 'overnight-boarding',
      isFetching: false,
      isSearchTypesLoaded: false,
      hasFailed: false,
      optionsByCategoryId: {},
      rankedCategories: [],
      inputs: {},
      fetchSearchOptions: jest.fn().mockName('fetchSearchOptions'),
      onSearchFieldChanged: jest.fn().mockName('onSearchFieldChanged'),
    };
  });

  describe('when mounting after search types have already been loaded', () => {
    beforeEach(() => {
      props = { ...props, isSearchTypesLoaded: true };
      rendered = shallow(<SearchOptions {...props} />);
    });

    it('fetches the search options for the current search type', () => {
      expect(props.fetchSearchOptions).toHaveBeenCalledWith(props.searchType);
    });
  });

  describe('when search types are not yet loaded', () => {
    beforeEach(() => {
      props = { ...props, isSearchTypesLoaded: false };
      rendered = shallow(<SearchOptions {...props} />);
    });

    describe('when search types then load', () => {
      beforeEach(() => {
        rendered.setProps({ isSearchTypesLoaded: true });
      });

      it('fetches the search options for the current search type', () => {
        expect(props.fetchSearchOptions).toHaveBeenCalledWith(props.searchType);
      });
    });
  });

  describe('when fetching', () => {
    beforeEach(() => {
      props = { ...props, isFetching: true };
      rendered = shallow(<SearchOptions {...props} />);
    });

    it('displays a loading spinner', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('when fetching failed', () => {
    beforeEach(() => {
      props = { ...props, isFetching: false, hasFailed: true };
      rendered = shallow(<SearchOptions {...props} />);
    });

    it('displays an error message with an option to reload', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('when fetching succeeded', () => {
    beforeEach(() => {
      const rankedCategories: OptionCategory[] = [
        {
          categoryId: 'service_details',
          rank: 1,
          displayName: 'Service Details',
        },
        {
          categoryId: 'service_details',
          rank: 2,
          displayName: 'Service Details',
        },
      ];
      props = {
        ...props,
        isSearchTypesLoaded: true,
        isFetching: false,
        hasFailed: false,
        rankedCategories,
      };
      rendered = shallow(<SearchOptions {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });

    describe('when the component updates but the search type remains the same', () => {
      beforeEach(() => {
        (props.fetchSearchOptions as jest.Mock).mockClear();
        rendered.setProps({ searchType: props.searchType });
      });

      it('does not refetch options for the same search type', () => {
        expect(props.fetchSearchOptions).not.toHaveBeenCalled();
      });
    });

    describe('when the search type changes to a new type', () => {
      beforeEach(() => {
        (props.fetchSearchOptions as jest.Mock).mockClear();
        rendered.setProps({ searchType: 'drop-in' });
      });

      it('refetches options for the new search type', () => {
        expect(props.fetchSearchOptions).toHaveBeenCalledWith('drop-in');
      });
    });
  });
});
