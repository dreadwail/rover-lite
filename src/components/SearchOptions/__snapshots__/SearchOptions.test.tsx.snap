// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOptions when fetching displays a loading spinner 1`] = `<Spinner />`;

exports[`SearchOptions when fetching failed displays an error message with an option to reload 1`] = `<ReloadableFailure />`;

exports[`SearchOptions when fetching succeeded renders correctly 1`] = `
<Fragment>
  <SearchOptionsCategory
    categoryId="service_details"
    displayName="Service Details"
    key="service_details"
    rank={1}
  >
    <SearchOptionsSet
      inputs={Object {}}
      onSearchFieldChanged={[MockFunction onSearchFieldChanged]}
    />
  </SearchOptionsCategory>
  <SearchOptionsCategory
    categoryId="service_details"
    displayName="Service Details"
    key="service_details"
    rank={2}
  >
    <SearchOptionsSet
      inputs={Object {}}
      onSearchFieldChanged={[MockFunction onSearchFieldChanged]}
    />
  </SearchOptionsCategory>
</Fragment>
`;
